src/bandersnatch/mirror.py

```diff
@@ -193,6 +115,144 @@ def _filter_packages(self) -> None:
                 else:
                     del self.packages_to_sync[package_name]
 
+    async def determine_packages_to_sync(self) -> None:
+        """
+        Update the self.packages_to_sync to contain packages that need to be
+        synced.
+        """
+        raise NotImplementedError()
+
+    async def package_syncer(self, idx: int) -> None:
+        logger.debug(f"Package syncer {idx} started for duty")
+        while True:
+            try:
+                package = self.package_queue.get_nowait()
+                await package.update_metadata(self.master, attempts=3)
+                await self.process_package(package)
+            except asyncio.QueueEmpty:
+                logger.debug(f"Package syncer {idx} emptied queue")
+                break
+            except PackageNotFound:
+                return  # I think this should be a continue
+            except Exception:
+                logger.exception(
+                    f"Error syncing package: {package.name}@{package.serial}"
+                )
+                self.errors = True
+
+            if self.errors and self.stop_on_error:
+                logger.error("Exiting early after error.")
+                sys.exit(1)
+
+    async def process_package(self, package: Package) -> None:
+        raise NotImplementedError()
+
+    async def sync_packages(self) -> None:
+        try:
+            self.package_queue: asyncio.Queue = asyncio.Queue()
+            # Sorting the packages alphabetically makes it more predictable:
+            # easier to debug and easier to follow in the logs.
+            for name in sorted(self.packages_to_sync):
+                serial = int(self.packages_to_sync[name])
+                await self.package_queue.put(Package(name, serial=serial))
+
+            sync_coros: List[Awaitable] = [
+                self.package_syncer(idx) for idx in range(self.workers)
+            ]
+            try:
+                await asyncio.gather(*sync_coros)
+            except KeyboardInterrupt:
+                # Setting self.errors to True to ensure we don't save Serial
+                # and thus save to disk that we've had a successful sync
+                self.errors = True
+                logger.info(
+                    "Cancelling, all downloads are forcibly stopped, data may be "
+                    + "corrupted. Serial will not be saved to disk. "
+                    + "Next sync will start from previous serial"
+                )
+        except (ValueError, TypeError):
+            # This is for when self.packages_to_sync isn't of type Dict[str, int]
+            self.errors = True
+
+    def finalize_sync(self) -> None:
+        return None
```

Comment:
I did this initially but actually I think we should force the user to implement it, even if it's just a no-op.  That way it'll be harder to accidentally misuse.